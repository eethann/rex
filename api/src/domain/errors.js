export class UserExistsError extends Error {}
export class UsernameRequiredError extends Error {}
