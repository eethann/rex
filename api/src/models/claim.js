import Sequelize from 'sequelize'

export default (sequelize, DataTypes) => {
  var Claim = sequelize.define('Claim', {
    id: {
      allowNull: false,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV1,
      type: DataTypes.UUID
    },
    url: DataTypes.TEXT,
    title: DataTypes.TEXT,
    imageUrl: DataTypes.TEXT,
    imageDescription: DataTypes.TEXT
  }, {tableName: 'claims'});
  Claim.associate = function(models) {
    Claim.belongsTo(models.User, {as: 'author'})
  };
  return Claim;
};
