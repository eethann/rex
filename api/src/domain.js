import {Claim, User, sequelize } from './models'

export const setupCleanDatabase = () => sequelize.sync({force: true})

export const clearDatabase = async () => {
  await Claim.destroy({truncate: true, cascade: true})
  await User.destroy({truncate: true, cascade: true})
}

export const cleanupDatabaseConnection = () => sequelize.close()
