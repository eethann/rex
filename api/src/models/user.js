import { UserExistsError, UsernameRequiredError } from '../domain/errors'

import securePassword from 'secure-password'
const pwd = securePassword({
  memlimit: securePassword.MEMLIMIT_DEFAULT,
  opslimit: securePassword.OPSLIMIT_DEFAULT
})


export default (sequelize, DataTypes) => {
  var User = sequelize.define('User', {
    id: {
      allowNull: false,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV1,
      type: DataTypes.UUID
    },
    username: DataTypes.TEXT,
    email: DataTypes.TEXT,
    password: DataTypes.BLOB
  }, {tableName: 'users'});
  User.associate = function(models) {
    User.hasMany(models.Claim, {foreignKey: 'authorID'})
  };
  User.generateHash = (password) =>
    pwd.hash(Buffer.from(password))

  User.findByUsername = (username) =>
    User.find({where: {username}})

  User.register = (user, password) => new Promise(async (resolve, reject) => {
    const { username } = user
    if (username) {
      const existingUser = await User.findByUsername(username)
      if (!existingUser) {
        return resolve(User.create({...user, password: await User.generateHash(password)}))
      } else {
        return reject(new UserExistsError(""))
      }
    } else {
      return reject(new UsernameRequiredError(""))
    }
  })

  return User;
};
