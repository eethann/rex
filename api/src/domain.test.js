import { User, sequelize } from './models'
import { UserExistsError, UsernameRequiredError } from './domain/errors'

import { setupCleanDatabase, clearDatabase, cleanupDatabaseConnection } from './domain'

import { travis } from './test/data'

beforeAll(setupCleanDatabase)

afterEach(clearDatabase)

afterAll(cleanupDatabaseConnection)

describe('user registration', () => {
  test('a new user can be created', async () => {
    const user = await User.register(travis, "kingme")
    expect(user.dataValues).toMatchSnapshot({
      id: expect.any(String),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      password: expect.anything()
    })
  })

  test('no username throws error', async () => {
    await expect(User.register({...travis, username: undefined}, "kingme")).rejects
      .toThrow(Error)
  })

  test('duplicate username throws error', async () => {
    await User.register({...travis, username: "travsquared"}, "kingme")
    await expect(User.register({...travis, username: "travsquared"}, "kingme")).rejects
      .toThrow(Error)
  })
})
