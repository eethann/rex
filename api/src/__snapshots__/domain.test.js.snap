// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user registration a new user can be created 1`] = `
Object {
  "createdAt": Any<Date>,
  "email": "travis@rexit.com",
  "id": Any<String>,
  "password": Anything,
  "updatedAt": Any<Date>,
  "username": "travis",
}
`;
